diagnostics/diagnostics_manager.rs

```diff
@@ -0,0 +1,217 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{
+    collections::HashSet,
+    error::Error,
+    fmt::format,
+    future::Future,
+    net::SocketAddr,
+    path::PathBuf,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    time::Instant,
+};
+
+use error::TypeDBError;
+use resource::constants::diagnostics::REPORTING_URI;
+use tonic::Status;
+use tonic_types::StatusExt;
+
+use crate::{
+    metrics::{ActionKind, DatabaseMetrics, LoadKind},
+    monitoring_server::MonitoringServer,
+    reporter::Reporter,
+    Diagnostics,
+};
+
+macro_rules! diagnostics_method {
+    ($(
+        pub fn $fn_name:ident(&self, $($arg_name:ident: $arg_type:ty),* $(,)?) ;
+    )*) => {
+        $(
+            pub fn $fn_name(&self, $($arg_name: $arg_type),*) {
+                self.diagnostics.$fn_name($($arg_name),*)
+            }
+        )*
+    };
+}
+
+#[derive(Debug)]
+pub struct DiagnosticsManager {
+    diagnostics: Arc<Diagnostics>,
+    reporter: Reporter,
+    monitoring_server: MonitoringServer,
+}
+
+impl DiagnosticsManager {
+    pub fn new(
+        deployment_id: String,
+        server_id: String,
+        distribution: String,
+        version: String,
+        data_directory: PathBuf,
+        monitoring_port: u16,
+        is_monitoring_enabled: bool,
+        is_reporting_enabled: bool,
+    ) -> Self {
+        let diagnostics = Arc::new(Diagnostics::new(
```

Comment:
`DiagnosticsManager` holds an `Arc` wrapping the only instance of `Diagnostics`. This `Arc` is shared between it, `MonitoringServer`, and `Reporter`.

Reply:
This ctor should take a `Diagnostics` value instead of passing a bunch of parameters through to it.

Reply:
From my point of view, it manages `Diagnostics` by itself, and it is responsible for creating `Diagnostics`. It's cleaner to avoid passing the arguments into multiple places (which can also lead to mistakes), but I'm not really convinced. 
Just like the `DatabaseManager`, which is responsible for **creating**, updating, deleting, and viewing databases. 

But I'm okay to give up if we believe that the argument passing is a strong argument, or you have something else to mention.

Reply:
> Just like the `DatabaseManager`, which is responsible for creating, updating, deleting, and viewing databases.
But when we're interacting with a database, that's still done via a `Database` value, access to which is provided through `DatabaseManager::database`, so it's not as if it's invisible.

My main argument here is that it will structure the arguments rather than having just a list of strings and numbers.

Reply:
Fine, I'll `Diagnostics` outside.

Reply:
It doesn't really help, actually. I either need to extract data from `diagnostics` for two params or:

```
impl DiagnosticsManager {
    pub fn new(
        deployment_id: String,
        diagnostics: Diagnostics,
        data_directory: PathBuf,
        monitoring_port: u16,
        is_monitoring_enabled: bool,
        is_reporting_enabled: bool,
    ) -> Self {
        let diagnostics = Arc::new(diagnostics);
        let reporter =
            Reporter::new(deployment_id, diagnostics.clone(), REPORTING_URI, data_directory, is_reporting_enabled);
        let monitoring_server = MonitoringServer::new(diagnostics.clone(), monitoring_port, is_monitoring_enabled);

        Self { diagnostics, reporter, monitoring_server }
    }
    }
    ```